### PR TITLE
Adding configuration file support.

### LINKED ISSUE #1: Add config file support
Suggestion: Use toml with https://github.com/alexcrichton/toml-rs

### PR DESCRIPTION
**Pull request by nthorne**
Add framework for configuring the application via configuration file.

Closes #1

### CODE REVIEW

File: src/cfg/mod.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+use std::io::Read;
+use toml::from_str;
+
+pub mod errors {
+    error_chain!{}
+}
+
+use self::errors::*;
+
+const CONFIGFILE: &'static str = "blacklung.cfg";
```
Comment (nthorne) [inline src/cfg/mod.rs]: Should perhaps be an argument, that defaults to this value.

File: src/cfg/mod.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+const USAGE: &'static str = "
+Blacklung server.
+
+Usage:
+    blacklung [--port=<port>]
+    blacklung (-h | --help)
+
+Options:
+    -h --help       Show this screen.
+    --port=<PORT>   Port to bind to. Needs to be detailed here or in configuration file.
```
Comment (nthorne) [inline src/cfg/mod.rs]: ... as an argument or a configuration item.

File: src/main.rs [hunk 3]
```diff
@@ -3,50 +3,43 @@
...
-
-#[derive(Debug, RustcDecodable)]
-struct Args {
-    flag_port: u16,
-}
 use blacklung::logging;
+use blacklung::cfg;
+use std::process;
+
+fn handle_error<T, V, E>(f: T) -> V
```
Comment (nthorne) [inline src/main.rs]: Rename to something more descriptive, and add documentation.

File: src/cfg/mod.rs (hunk 1 above)
Comment (tripokey) [reply src/cfg/mod.rs]: Yes, we should have an internal default value which can be overridden with a cfg file value which in turn can be overridden with a command line argument.

File: src/cfg/mod.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+struct Args {
+    flag_port: Option<u16>,
+}
+
+/// This struct holds data that can be read from the config file;
+/// hence the Option.
+#[derive(Debug, Deserialize)]
+struct ConfigurationFile {
+    pub port: Option<u16>,
+}
```
Comment (tripokey) [inline src/cfg/mod.rs]: type PortType = Option<u16>;

and then use PortType in the Args and ConfigurationFile structs.

File: src/cfg/mod.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    // default constructed ConfigurationFile in that case..
+    match File::open(CONFIGFILE) {
+        Ok(mut file) => {
+            let mut data = String::new();
+            // .. but if it does exist, then reading should succeed..
+            file.read_to_string(&mut data).
+                chain_err(|| "Failed to read from file")?;
+
+            // .. and the parsing should succeed..
+            let cfg: ConfigurationFile = from_str(data.as_str()).
```
Comment (tripokey) [inline src/cfg/mod.rs]: data.as_str() can be replaced with &data if desired

File: src/cfg/mod.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            Ok(cfg)
+        },
+        Err(_) => Ok(ConfigurationFile::new())
+    }
+}
+
+fn join(cfg: ConfigurationFile, args: Args) -> Result<Config> {
+    let _port = match cfg.port.or(args.flag_port) {
+        Some(v) => v,
+        None    => bail!("Port not configured or given as command line argument."),
```
Comment (tripokey) [inline src/cfg/mod.rs]: Use hardcoded default value instead

File: src/main.rs (hunk 3 above)
Comment (tripokey) [reply src/main.rs]: Not anything we can use now but just for reference... https://internals.rust-lang.org/t/rfc-mentoring-opportunity-permit-in-main/4600

**Review by tripokey — CHANGES_REQUESTED** [review]

File: src/cfg/mod.rs (hunk 1 above)
Comment (nthorne) [reply src/cfg/mod.rs]: Done (for port)

File: src/cfg/mod.rs (hunk 2 above)
Comment (nthorne) [reply src/cfg/mod.rs]: Done

File: src/cfg/mod.rs (hunk 4 above)
Comment (nthorne) [reply src/cfg/mod.rs]: Good idea. Done.

File: src/cfg/mod.rs (hunk 5 above)
Comment (nthorne) [reply src/cfg/mod.rs]: Done.

File: src/cfg/mod.rs (hunk 6 above)
Comment (nthorne) [reply src/cfg/mod.rs]: Makes sense. Done.

**Review by tripokey — APPROVED** [review]

**Pull request merged**